### PR TITLE
Drop examples with no completion left after truncation in DPO and KTO

### PR DESCRIPTION
**Pull request by qgallouedec**
When truncation removes every completion token from an example, it carries no learning signal (its completion mask becomes all zeros) but still consumes compute. This happens with `truncation_mode="keep_start"` when the prompt alone already fills `max_length`. This PR drops such examples during dataset preparation.

Mirrors the same behavior added for SFT in #6320

## How

DPO and KTO truncate in the **collator** (per batch), so there's no post-truncation labels column to inspect at prep time. Instead, a `.filter` after tokenization drops examples whose prompt alone reaches `max_length` (`len(prompt_ids) >= max_length`) — exactly the condition under which `keep_start` truncation erases the whole completion.

- **DPO**: applied only for `truncation_mode="keep_start"`. `keep_end` keeps the completion end, so nothing is dropped there. The prompt is shared by chosen and rejected, so the single condition covers both.
- **KTO**: the collator always truncates `keep_start`, so the filter always applies when `max_length is not None`.

Both work for `Dataset` and `IterableDataset`. For DPO/KTO vision datasets, dataset preparation is skipped (truncation stays in the collator), so this only affects the text path.

## Example

```python
from datasets import Dataset
from trl import DPOConfig, DPOTrainer

ds = Dataset.from_dict({
    "prompt": ["Hi", "This is a very long prompt that fills up the whole max_length budget on its own"],
    "chosen": [" there", " yes"],
    "rejected": [" bye", " no"],
})

trainer = DPOTrainer(
    "trl-internal-testing/tiny-Qwen2ForCausalLM-2.5",
    args=DPOConfig(max_length=6),
    train_dataset=ds,
)

len(trainer.train_dataset)  # 1 — the long-prompt example is dropped (before: 2)
```

---

> [!NOTE]
> **Low Risk**
> Dataset filtering during prep only; no change to loss math or model forward. Risk is silently shrinking train/eval sets if many prompts hit max_length.
> 
> **Overview**
> **DPO** and **KTO** now remove training examples during dataset prep when the tokenized prompt alone reaches `max_length`, so collator truncation would erase every completion token and leave no learning signal (mirroring SFT #6320).
> 
> After tokenization, a `.filter` keeps only rows with `len(prompt_ids) < max_length`. **DPO** applies this only when `truncation_mode="keep_start"`; with `keep_end`, completions are preserved so nothing is dropped. **KTO** always uses `keep_start` in the collator, so the filter runs whenever `max_length` is set. Text-only prep paths are affected; vision datasets still truncate only in the collator.
> 
> Regression tests in `test_dpo_trainer.py` and `test_kto_trainer.py` assert that a long-prompt row is dropped and the short one remains.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3df5a3183e6e8ed81d1c6c0f5765231d44e99bb7. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -1035,6 +1035,14 @@ def tokenize_fn(example, processing_class):
                 **map_kwargs,
             )
 
+            # Drop examples whose prompt alone fills `max_length`: with `keep_start` truncation the collator would
+            # remove every completion token, leaving no learning signal. `keep_end` keeps the completion end, so
+            # nothing is dropped there.
+            if args.max_length is not None and args.truncation_mode == "keep_start":
+                if isinstance(dataset, Dataset):  # `IterableDataset.filter` does not support `desc`
+                    map_kwargs["desc"] = f"Dropping examples with fully truncated completion from {dataset_name} dataset"
+                dataset = dataset.filter(lambda example: len(example["prompt_ids"]) < args.max_length, **map_kwargs)
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/dpo_trainer.py:1058]: The filter keys off `args.truncation_mode`, which is the documented contract for truncation policy, a custom collator is expected to respect it. Gating this on collator identity would mean threading the collator into dataset prep to cover a case that doesn't exist today, which isn't a tradeoff worth making here.

If you pass a collator that ignores `args.max_length`, leave `truncation_mode="keep_end"` and nothing is dropped.

File: trl/trainer/kto_trainer.py [hunk 2]
```diff
@@ -1080,6 +1080,14 @@ def tokenize_fn(example, processing_class):
 
             dataset = dataset.map(tokenize_fn, fn_kwargs={"processing_class": processing_class}, **map_kwargs)
 
+            # Drop examples whose prompt alone fills `max_length`: with `keep_start` truncation the collator would
+            # remove every completion token, leaving no learning signal. `keep_end` keeps the completion end, so
+            # nothing is dropped there.
+            if args.max_length is not None:
+                if isinstance(dataset, Dataset):  # `IterableDataset.filter` does not support `desc`
+                    map_kwargs["desc"] = f"Dropping examples with fully truncated completion from {dataset_name} dataset"
+                dataset = dataset.filter(lambda example: len(example["prompt_ids"]) < args.max_length, **map_kwargs)
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/kto_trainer.py:1073]: Same as the DPO thread: KTO's collator always truncates `keep_start`, so the filter matches the trainer's actual policy. A custom collator with a different policy is out of scope for now, happy to revisit if it comes up.

**Comment by qgallouedec** [comment]
merging this as it's been open for a while
